pkg/registry/scheduling/rest/storage_scheduling.go

```diff
@@ -89,51 +94,112 @@ func (p RESTStorageProvider) v1Storage(apiResourceConfigSource serverstorage.API
-					if apierrors.IsNotFound(err) {
-						// TODO: Remove this explicit conversion after scheduling api move to v1
-						v1beta1PriorityClass := &schedulingv1beta1.PriorityClass{}
-						if err := schedulingapiv1beta1.Convert_scheduling_PriorityClass_To_v1beta1_PriorityClass(pc, v1beta1PriorityClass, nil); err != nil {
-							return false, err
-						}
-						_, err := schedClientSet.PriorityClasses().Create(v1beta1PriorityClass)
-						if err != nil && !apierrors.IsAlreadyExists(err) {
-							return false, err
-						} else {
-							klog.Infof("created PriorityClass %s with value %v", pc.Name, pc.Value)
-						}
+		for _, pc := range scheduling.SystemPriorityClasses() {
+			_, err := schedClientSet.PriorityClasses().Get(pc.Name, metav1.GetOptions{})
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					// TODO: Remove this explicit conversion after scheduling api move to v1
+					v1beta1PriorityClass := &schedulingv1beta1.PriorityClass{}
+					if err := schedulingapiv1beta1.Convert_scheduling_PriorityClass_To_v1beta1_PriorityClass(pc, v1beta1PriorityClass, nil); err != nil {
+						return false, err
+					}
+					_, err := schedClientSet.PriorityClasses().Create(v1beta1PriorityClass)
+					if err != nil && !apierrors.IsAlreadyExists(err) {
+						return false, err
 					} else {
-						// Unable to get the priority class for reasons other than "not found".
-						klog.Warningf("unable to get PriorityClass %v: %v. Retrying...", pc.Name, err)
-						return false, nil
+						klog.Infof("created PriorityClass %s with value %v", pc.Name, pc.Value)
 					}
+				} else {
+					// Unable to get the priority class for reasons other than "not found".
+					klog.Warningf("unable to get PriorityClass %v: %v. Retrying...", pc.Name, err)
+					return false, nil
 				}
 			}
-			klog.Infof("all system priority classes are created successfully or already exist.")
-			return true, nil
-		})
-		// if we're never able to make it through initialization, kill the API server.
+		}
+		klog.Infof("all system priority classes are created successfully or already exist.")
+		return true, nil
+	})
+}
+
+// addDefaultSystemQuota adds a quota which allows unlimited critical pods to be created in `kube-system` namespace
+// This quota can be deleted/updated by cluster-admin. If the cluster-admin wants critical pods to be created in
+// namespace other than `kube-system`, he/she can configure quotas which allows these critical pods to be created
+// in that namespace.
+func addDefaultSystemQuota(hookContext genericapiserver.PostStartHookContext) error {
+	return wait.Poll(1*time.Second, 30*time.Second, func() (done bool, err error) {
+		client := clientset.NewForConfigOrDie(hookContext.LoopbackClientConfig)
+		_, err = client.CoreV1().ResourceQuotas(metav1.NamespaceSystem).Get(DefaultSystemQuota, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			_, err := client.CoreV1().ResourceQuotas(metav1.NamespaceSystem).Create(systemDefaultQuota())
```

Comment:
doesnt this prevent the user from deleting the quota if its just recreated?

Reply:
I don't think so, the re-creation happens during bootstrap stage where user cannot interact with the api-server yet. Once the bootstrap is complete, users will be able to delete the quota.

Reply:
> doesnt this prevent the user from deleting the quota if its just recreated?

but it will come back on the next server restart.

Reply:
@deads2k  do we have an option other than documenting the behavior, and that an operator can only update the default quota not delete it?